ONSdigital/census-contact-centre-service
# Cr 187

# Motivation and Context

This change is required because of CR-187, which states:

POST /cases/
{caseid}
/fulfilment/sms
Apart from the slight differences in the request object sent to this endpoint, this is largely identical in functionality to the fulfilment by POST endpoint, so should not take long to implement, as it will be largely reuse, with the obvious difference being in the deliveryChannel check must be for SMS.


# What has changed
	•	The CaseServiceImpl class has changed to include an implementation of the fulfilmentRequestBySMS method. 
	•	Three unit tests have been added to the CaseServiceImplTest class (two happy path scenarios and one failure scenario).
	•	The fulfilment/post and fulfilment/sms endpoint methods have had an if statement added, to each of them, to make sure that the caseid in the rest endpoint matches the one in the request body.
	•	A couple of other junit tests, which were affected by the two endpoint changes, have been amended as necessary (to avoid failures). 
 
The fulfilmentRequestBySMS method will receive an incoming SMSFulfilmentRequestDTO, from the Contact Centre, then use it to create an outgoing FulfilmentRequestedEvent, to send to RM. The FulfilmentRequestedEvent gets put onto the events exchange, which in turn puts it onto an outgoing queue.

The incoming JSON, for the SMSFulfilmentRequestDTO, would look something like this:

{
  "caseId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "telNo": "+447890000000",
  "fulfilmentCode": "UACHHT1",
  "dateTime": "2016-11-09T11:44:44.797Z"
}

The outgoing JSON, for the FulfilmentRequestedEvent, would look something like this:

{"event":{"type":"FULFILMENT_REQUESTED","source":"CONTACT_CENTRE_API","channel":"CC","dateTime":"2019-05-30T16:23:58.782Z","transactionId":"3f4c0f72-e98c-4f21-b9c7-22fd7da42c8b"},"payload":{"fulfilmentRequest":{"fulfilmentCode":"UACHHT1","caseId":"3fa85f64-5717-4562-b3fc-2c963f66afa6","individualCaseId":null,"address":{"addressLine1":null,"addressLine2":null,"addressLine3":null,"townName":null,"postcode":null,"region":null,"latitude":null,"longitude":null,"uprn":null,"arid":null,"addressType":null,"estabType":null},"contact":{"title":null,"forename":null,"surname":null,"email":null,"telNo":"+447890000000"}}}}






# How to test?

Firstly, if you have not already done so, you need to create a queue, manually, using the RabbitMQ management console NB. name it something like ‘Fulfilment.Test’ and manually bind it to the events exchange using the following routing key: event.fulfilment.request

Then do as follows:  Add Peter’s class, named CaseServiceFakeStub into the following package:
uk.gov.ons.ctp.integration.contactcentresvc.endpoint

Amend the application.yml so that it contains the following for the case service settings: case-service-settings: case-by-id-query-path: /cases-rm-fake/{case-id} case-by-uprn-query-path: /cases-rm-fake/uprn/{uprn} case-by-case-reference-query-path: /cases-rm-fake/ref/{reference} rest-client-config: username: serco_cks password: temporary scheme: http host: localhost port: 8171

Run the following command to launch RabbitMQ from a terminal (in the census-contact-centre-service repo):
Docker-compose up -d

Save, compile and run the Contact Centre service.

Run Postman and use the example JSON above to post an SMSFulfilmentRequestDTO  to the following endpoint:
http://localhost:8171/cases/3fa85f64-5717-4562-b3fc-2c963f66afa6/fulfilment/sms
NB. It is important to make sure that the caseid in the above URL is the same as the incoming JSON, otherwise you should receive a ‘BAD REQUEST’ error. If all works correctly then a FulfilmentRequestedEvent message (I.e. the outgoing JSON) should appear in the queue that you created earlier.
